File: src/CorporateMessengerWithAgile/ReactApp/reactapp.client/src/controllers/TaskController.ts
// src/controllers/TaskController.ts
import { AuthenticatedController } from './abstract/AuthenticatedController';
import {
    type Guid, Result,
    TaskItemSummaryDtoSchema,
    TaskItemWithStatusDtoSchema,
    type TaskItemSummaryDto,
    type TaskItemWithStatusDto
} from '../models';
import { convertToArray, validateWithSchema } from '../utils/validation';
import { AuthController } from '.';

export class TaskController extends AuthenticatedController {
    constructor(authController: AuthController) { super(authController, '/Task'); }

    async getTaskItemsBySprintWithStatus(sprintId: Guid): Promise<Result<TaskItemWithStatusDto[]>> {
        const result = await this.request('GET', `/sprint/${sprintId}/task-items-with-status`);
        if (result.isFailure) return result as Result<TaskItemWithStatusDto[]>;
        return convertToArray<TaskItemWithStatusDto>(TaskItemWithStatusDtoSchema, result.value);
    }

    async getTaskItemsBySprint(sprintId: Guid): Promise<Result<TaskItemSummaryDto[]>> {
        const result = await this.request('GET', `/sprint/${sprintId}/task-items`);
        if (result.isFailure) return result as Result<TaskItemSummaryDto[]>;
        return convertToArray<TaskItemSummaryDto>(TaskItemSummaryDtoSchema, result.value);
    }

    async getTaskItemsByProject(projectId: Guid): Promise<Result<TaskItemSummaryDto[]>> {
        const result = await this.request('GET', `/project/${projectId}`);
        if (result.isFailure) return result as Result<TaskItemSummaryDto[]>;
        return convertToArray<TaskItemSummaryDto>(TaskItemSummaryDtoSchema, result.value);
    }

    async createTaskItem(data: { title: string; description: string; priority: number; complexity: number; deadline: string; projectId: Guid; authorId: Guid; responsibleId: Guid; sprintWithLastMentionId?: Guid; parentTaskId?: Guid; }): Promise<Result<TaskItemSummaryDto>> {
        const result = await this.request('POST', '', data);
        if (result.isFailure) return result as Result<TaskItemSummaryDto>;
        return validateWithSchema(TaskItemSummaryDtoSchema, result.value);
    }
}

File: src/CorporateMessengerWithAgile/ReactApp/ReactApp.Server/Controllers/TaskController.cs
using Application.Dto;
using Application.Dto.Summary;
using Application.Entity.TaskItems.Commands.TaskItemCreate;
using Application.Entity.TaskItems.Queries.TaskItemsGetByProject;
using Application.Entity.TaskItems.Queries.TaskItemsGetBySprint;
using Application.Entity.TaskItems.Queries.TaskItemsGetBySprintWithStatus;
using Domain.Result;
using MediatR;
using Microsoft.AspNetCore.Mvc;
using ReactApp.Server.Controllers.Abstract;

namespace ReactApp.Server.Controllers
{
    [Route("cmwa/[controller]")]
    public class TaskController(ISender sender) : AuthorizedBaseController(sender)
    {
        [HttpGet("sprint/{sprintId}/task-items-with-status")]
        public async Task<IEnumerable<TaskItemWithStatusDto>> GetTaskItemsBySprintWithStatus(
            [FromRoute] Guid sprintId,
            CancellationToken cancellationToken = default
        ) => await SendAuth<TaskItemsGetBySprintWithStatusQuery, IEnumerable<TaskItemWithStatusDto>>(new TaskItemsGetBySprintWithStatusQuery(sprintId), cancellationToken);

        [HttpGet("sprint/{sprintId}/task-items")]
        public async Task<IEnumerable<TaskItemSummaryDto>> GetTaskItemsBySprint(
            [FromRoute] Guid sprintId,
            CancellationToken cancellationToken = default
        ) => await SendAuth<TaskItemsGetBySprintQuery, IEnumerable<TaskItemSummaryDto>>(new TaskItemsGetBySprintQuery(sprintId), cancellationToken);

        [HttpGet("project/{projectId}")]
        public async Task<IEnumerable<TaskItemSummaryDto>> GetTaskItemsByProject(
            [FromRoute] Guid projectId,
            CancellationToken cancellationToken = default
        ) => await SendAuth<TaskItemsGetByProjectQuery, IEnumerable<TaskItemSummaryDto>>(new TaskItemsGetByProjectQuery(projectId), cancellationToken);

        [HttpPost]
        public async Task<ActionResult<TaskItemSummaryDto>> CreateTaskItem(
            [FromBody] CommandCreateTaskItem request,
            CancellationToken cancellationToken = default
        ) => (await SendAuth<CommandCreateTaskItem, Result<TaskItemSummaryDto>>(request, cancellationToken)).ToActionResult();
    }
}
